### ISSUE TITLE
Lint/UselessAssignment incorrectly detected variable to remove

### ISSUE THREAD

**Issue by janko** [rubocop/rubocop#14520]
## Expected behavior

I expected Rubocop to not discard legitimate variable assignment.

## Actual behavior

Rubocop's `Lint/UselessAssignment` classified variable assignment as useless, and even offered to remove it via "safe" autocorrection.

## Steps to reproduce the problem

```bash
$ RUBOCOP_TARGET_RUBY_VERSION=3.4 rubocop --display-only-safe-correctable --debug lint.rb
```
```rb
# lint.rb

def generate_pin
  tried_codes = Set.new
  i = 0
  try = 0

  # possible_code_combinations represents the theoretical maximum number of
  #   pins of the desired length
  # max_tries represents a sane number after which to stop trying to
  #   generate a pin
  #
  # Both are required as `i` (which counts up to possible_code_combinations)
  # increments only when a pin is tried for the first time. While `try` is
  # incremented in every iteration
  while i < possible_code_combinations && try < max_tries
    try += 1
    pin.code = generate_code_candidate

    # Populate validaton errors on the model
    pin.valid?

    # Skip codes that we've already tried
    # http://ruby-doc.org/stdlib-2.6.3/libdoc/set/rdoc/Set.html#method-i-add-3F
    next if tried_codes.add?(pin.code).nil? || weak?

    pin.save if persist

    if persist && pin.invalid? && available?
      # The pin isn't taken or weak but it's invalid - something went wrong
      raise(ActiveRecord::RecordInvalid, pin)
    elsif available?
      return pin
    end

    i += 1
    try = 0
  end

  # the loop exited without returning a pin - there are no more valid codes
  raise(CombinationPoolExhaustedError)
end
```
```
For /Users/janko/Code/temp: Default configuration from /Users/janko/.rbenv/versions/3.4.4/lib/ruby/gems/3.4.0/gems/rubocop-1.80.2/config/default.yml
Inspecting 1 file
Scanning /Users/janko/Code/temp/lint.rb
Loading cache from /Users/janko/.cache/rubocop_cache/f82a208541caa928e4be545413e42cc2efe44899/54ad3a583ed8da5bef3737984da3debc1188c038/423be99960a6ed0b884a2847303937270ecd19c0
W

Offenses:

lint.rb:15:5: W: [Correctable] Lint/UselessAssignment: Useless assignment to variable - try.
    try += 1
    ^^^

1 file inspected, 1 offense detected, 1 offense autocorrectable
Finished in 0.0411769999191165 seconds
```

## RuboCop version

```
$ RUBOCOP_TARGET_RUBY_VERSION=3.4 rubocop -V
1.80.2 (using Parser 3.3.8.0, Prism 1.4.0, default, rubocop-ast 1.46.0, analyzing as Ruby 3.4, running on ruby 3.4.4) [arm64-darwin24]
```
